### PR TITLE
test(core): e2e test cases core maciState/Poll

### PR DESCRIPTION
**Pull request by baumstern**
This PR adds end-to-end test cases of the off-chain state machine, focusing on sanity checks.

Related to #792

### CODE REVIEW

**Review by ctrlc03 — CHANGES_REQUESTED** [review]
Thanks for this, it can be very useful to ensure MACI works correctly, though I have some thoughts:

Currently these are gearing towards being e2e tests for the MaciState/Poll classes, which is definitely useful, though I wonder if it would be best to focus on unit testing MaciState/Poll as we already have e2e/integration tests that touch on MACI from 0 to end (though what we have now as e2e and integration tests are not the best, and in general they do not "test" much as failures could be for a variety of reasons - we have an issue open to track this #869).

If you strongly feel to keep these as it is, then I would ask you to please rename everything appropriately, e.g. move to a e2e folder inside  core tests, rename to not be processMessages specific (as you are relying on tally votes too). 

Otherwise, we now have an open [PR](https://github.com/privacy-scaling-explorations/maci/pull/914) which refactors the `processMessage` related functions, making `processMessage` public and parameterised with custom errors. This way it is now much easier to do unit testing and actually test that the different results are because of the different conditions (which is what is missing in this PR as well as our e2e/integration tests suite and that we need to fix). - Also the TestHarness class can possibly be adapted to work on our integration tests

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts:178]: The loop generates a variety of nonces to simulate different incorrect values

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts:218]: right, it's meant to ensure messages has published via `publish()` but ignored during `processMessages()`

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts:225]: > why the message wasn't processed
due to an incorrect state tree index

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts:249]: due to an incorrect signature

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts:292]: it's due to an invalid coordinator key

**Comment by baumstern** [comment]
> Thanks for this, it can be very useful to ensure MACI works correctly, though I have some thoughts:
> 
> Currently these are gearing towards being e2e tests for the MaciState/Poll classes, which is definitely useful, though I wonder if it would be best to focus on unit testing MaciState/Poll as we already have e2e/integration tests that touch on MACI from 0 to end (though what we have now as e2e and integration tests are not the best, and in general they do not "test" much as failures could be for a variety of reasons - we have an issue open to track this #869).
> 
> If you strongly feel to keep these as it is, then I would ask you to please rename everything appropriately, e.g. move to a e2e folder inside core tests, rename to not be processMessages specific (as you are relying on tally votes too).
> 
> Otherwise, we now have an open [PR](https://github.com/privacy-scaling-explorations/maci/pull/914) which refactors the `processMessage` related functions, making `processMessage` public and parameterised with custom errors. This way it is now much easier to do unit testing and actually test that the different results are because of the different conditions (which is what is missing in this PR as well as our e2e/integration tests suite and that we need to fix). - Also the TestHarness class can possibly be adapted to work on our integration tests

These are e2e tests at the `maci-core` package level. They don't cover inter-package level e2e testing, which involve both onchain and zk components. The main objective here is to ensure the integrity of the `maci-core` package's module, with a specific focus on the behavior of `processMessages()`. This is something our other end-to-end/integration tests don't cover and is outside their scope.

As a side note, these tests are black-box in nature, which is why they rely on the results of `tallyVote()`

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts]: This is specifically designed for testing purposes and does not require exporting

Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts]: Besides this point, we can consider designing a high-level module similar to this, something like:

```
class MACIHighLevel {
  init();
  vote();
  tally();
}
```
However, this is beyond the scope of the current PR

**Comment by baumstern** [comment]
Thanks for the feedback @ctrlc03 , I've changed file name and modified it to check for a specific throw condition. Please have another look

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (ctrlc03) [reply core/ts/__tests__/ProcessMessage.test.ts]: sure, the text says that, but where is the guarantee that it wasn't processed because of that?

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (ctrlc03) [reply core/ts/__tests__/ProcessMessage.test.ts]: The TestHarness class is still without any comments/documentation

Comment (ctrlc03) [reply core/ts/__tests__/ProcessMessage.test.ts]: so we can't re-use it other tests? 😔

Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts]: actually, TestHarness has been liberated for the good of all tests in 687bc92

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts]: the test cases here follow [black-box testing](https://csrc.nist.gov/glossary/term/black_box_testing). We establish a baseline with `processMessages() `processing a valid message, then in isolation, we alter one variable at a time—nonce, state tree index, vote weight—to observe specific failure conditions. This technique ensures each failure is attributable to its respective variable change

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (ctrlc03) [reply core/ts/__tests__/ProcessMessage.test.ts]: 🙏  thanks for that and for documenting it

File: core/ts/__tests__/ProcessMessage.test.ts
Comment (ctrlc03) [reply core/ts/__tests__/ProcessMessage.test.ts]: Cool, thanks for explaining, now it is more clear what you intended to do. I ask you that you please document this properly (you can even add a section to the testing doc inside the website), create own folder (as currently they are in the same place as where unit tests are supposed to be, based on how we have been organizing other TS packages) and ensure appropriate naming (smth like MaciTest.e2e or whatnot)

Comment (baumstern) [reply core/ts/__tests__/ProcessMessage.test.ts]: @ctrlc03 just to clarify, we're already on the same page with the testing approach – it's consistent with what we've got in MaciState.test.ts. No extra steps needed! The link I included is just for reference: https://github.com/privacy-scaling-explorations/maci/blob/dev/core/ts/__tests__/MaciState.test.ts#L613-L628

**Comment by baumstern** [comment]
Hey @ctrlc03, I need write access to push some updates. I've fixed the ESLint errors on the branch

**Comment by ctrlc03** [comment]
> Hey @ctrlc03, I need write access to push some updates. I've fixed the ESLint errors on the branch

Sorry for taking this long to get back to you! Would you be happy for us to complete this: we would fix conflicts and push a commit to fix eslint errors. Otherwise you could open a PR from a fork? up to you and how you prefer to get this done 🙂

**Comment by baumstern** [comment]
> > Hey @ctrlc03, I need write access to push some updates. I've fixed the ESLint errors on the branch
> 
> Sorry for taking this long to get back to you! Would you be happy for us to complete this: we would fix conflicts and push a commit to fix eslint errors. Otherwise you could open a PR from a fork? up to you and how you prefer to get this done 🙂

Check out the latest fixes [here](https://github.com/baumstern/maci/commits/coretest/), and you can go from there. I would like to avoid opening redundant PRs and keep things tidy

**Comment by ctrlc03** [comment]
> > > Hey @ctrlc03, I need write access to push some updates. I've fixed the ESLint errors on the branch
> > 
> > 
> > Sorry for taking this long to get back to you! Would you be happy for us to complete this: we would fix conflicts and push a commit to fix eslint errors. Otherwise you could open a PR from a fork? up to you and how you prefer to get this done 🙂
> 
> Check out the latest fixes [here](https://github.com/baumstern/maci/commits/coretest/), and you can go from there. I would like to avoid opening redundant PR and keep things tidy

Updated, thanks for your contribution :) lmk if it looks good to you

File: core/ts/__tests__/utils.ts
Comment (baumstern) [inline core/ts/__tests__/utils.ts:153]: we should decide whether to keep this one or use the other in `core/ts/__tests__/utils/MaciTestHarness.ts`

**Comment by baumstern** [comment]
> > > > Hey @ctrlc03, I need write access to push some updates. I've fixed the ESLint errors on the branch
> > > 
> > > 
> > > Sorry for taking this long to get back to you! Would you be happy for us to complete this: we would fix conflicts and push a commit to fix eslint errors. Otherwise you could open a PR from a fork? up to you and how you prefer to get this done 🙂
> > 
> > 
> > Check out the latest fixes [here](https://github.com/baumstern/maci/commits/coretest/), and you can go from there. I would like to avoid opening redundant PR and keep things tidy
> 
> Updated, thanks for your contribution :) lmk if it looks good to you Docs explaining different test suites in core: https://maci.pse.dev/docs/testing-in-details

thanks, the change looks good overall. just left one small comment

File: core/ts/__tests__/utils.ts
Comment (ctrlc03) [reply core/ts/__tests__/utils.ts]: oh right my bad I forgot to remove one of them, good catch. Kept inside test utils

Comment (baumstern) [reply core/ts/__tests__/utils.ts]: > oh right my bad I forgot to remove one of them, good catch. Kept inside test utils

cool, looks good to me!

**Review by ctrlc03 — APPROVED** [review]

**Pull request merged**
